File: examples/chapter6/table6-2.R
rm(list=ls())
library('pls')
library(readr)
library(Renvlp) 
library(matrixcalc)

###################
# Skagerberg data #
###################

#### PPLS PARTIAL OLS and ENV partial table 6.2 

#### Columns 2, 3, 4

aux <- read_csv("data/chapter6/Process_data_Skagerberg1992 - Sheet1.csv")

# For reduction
X1=as.matrix(aux[,1:20])

# No reduction
X2=as.matrix(aux[,21:22])

Y=as.matrix(log(aux[,23:28]))
Y=Y%*%diag(diag(cov(Y))^(-1/2))

aux$Y=Y


 
Dmax=18
 
n=nrow(aux)
pls_predic=array(NaN,dim=c(n,Dmax,6) )
ols_predic=array(NaN,dim=c(n,Dmax,6) )
env_predic=array(NaN,dim=c(n,Dmax,6) )
 


####PPLS PARTIAL OLS and ENV partial
for (d in 1:Dmax){
  for (i in 1:n){
    
    # training
    aux1=aux[-i,]
    
    # testing
    aux2=aux[i,]
    
    ### Matrix X to reduce
    AA=as.matrix(aux1[,1:20],ncol=20)
  
    ### Regression of the continuous X that we do not reduce
    A=lm(AA~Tw+S,data=aux1)
    
    # Save residuals
    aux1$SS=A$residuals 
    
    ### Predict on new data
    A_1=predict(A,newdata=aux2)
    
    # Center response in training 
    aux1$ynew=scale(aux1$Y,scale=FALSE)
    
    ## Do PLS on the training residuals
    m=plsr(ynew~SS,ncomp=d,data=aux1)
    
    # Perform regression on non-reduced variables
    j=lm(ynew~Tw+S,data=aux1)
    
    # Predict the new datapoint from non-reduced model
    j_2=predict(j,newdata=aux2)
    
    # Final prediction on new data
    yhat=apply(aux1$Y,2,mean)+j_2+as.numeric(as.matrix((aux2[,1:20]-A_1),nrow=1)%*%as.matrix(m$coefficients[,1:6,d],ncol=6))

    # Save PLS prediction
    pls_predic[i,d,]=yhat
    
    a=lm(Y~T1+T2+T3+T4+T5+T6+T7+T8+T9+T10+T11+T12+T13+T14+T15+T16+T17+T18+T19+T20+Tw+S
         ,data=aux1)
    
    # Save OLS prediction
    ols_predic[i,d,]=predict(a,newdata=aux2)
    
    M=eppls(X1[-i,],X2[-i,],Y[-i,],d)
    
    # Save Envelope prediction
    env_predic[i,d,]=t(M$beta1)%*%(X1[i,]-M$mu1)+t(M$beta2)%*%(X2[i,]-M$mu2)+M$muY
    
  }
}


p.ols=NULL
p.pls=NULL
p.env=NULL

for (d in 1:Dmax){
   p.ols[d]= mean(sqrt(apply((aux$Y-ols_predic[,d,])^2,1,mean)))
   p.pls[d]=mean(sqrt(apply((aux$Y-pls_predic[,d,])^2,1,mean)))
   p.env[d]=mean(sqrt(apply((aux$Y-env_predic[,d,])^2,1,mean)))
}


##table 6.2 columns 2, 3,4

table6_2 <- matrix(c(1:15), ncol=5, byrow=TRUE)

table6_2[1,] <- c('OLS', 'PLS (P.Predictor)', 'Envelope (P.Predictor)', 'PLS (P.Response)', 'Envelope (P.Response)')
rownames(table6_2) <- c('Method', 'No. components', 'predRMSE')

table6_2[3,1] = p.ols[1]
table6_2[2,1] = "p1 = 20"
table6_2[2,2] = which.min(p.pls)
table6_2[3,2] = p.pls[which.min(p.pls)]
table6_2[2,3] = which.min(p.env)
table6_2[3,3] = p.env[which.min(p.env)]

#######

#########COLUMNS 5,6 

# For reduction
X1=as.matrix(aux[,1:20])

# No reduction
X2=as.matrix(aux[,21:22])


Y=as.matrix(log(aux[,23:28]))
Y=Y%*%diag(diag(cov(Y))^(-1/2))

aux$Y=Y

fit_penv=array(NaN, dim=c(nrow(X1),ncol(Y),ncol(X1)))
error_penv=array(NaN, dim=c(nrow(X1),ncol(Y),ncol(X1)))

fit_prpls=array(NaN, dim=c(nrow(X1),ncol(Y),ncol(X1)))
error_prpls=array(NaN, dim=c(nrow(X1),ncol(Y),ncol(X1)))


##### Do envelope for different d
for (d in 1:ncol(Y)){
 for (i in 1:nrow(X1)){
   fit <- penv(X1[-i,],X2[-i,],Y[-i,], d)
   
   fit_penv[i,,d]=colMeans(Y[-i,])+(X1[i,]-colMeans(X1[-i,]))%*%t(fit$beta1)+(X2[i,]-colMeans(X2[-i,]))%*%t(fit$beta2)
   error_penv[i,,d]=(Y[i,]-colMeans(Y[-i,])-(X1[i,]-colMeans(X1[-i,]))%*%t(fit$beta1)-(X2[i,]-colMeans(X2[-i,]))%*%t(fit$beta2))^2
   
 }
}


#### PPLS, for d=1 we do it separately

for (i in 1:nrow(X1)){
 X1c=scale(X1[-i,],scale=FALSE)
 X2c=scale(X2[-i,],scale=FALSE)
 
 resY2=as.matrix(lm(Y[-i,]~X2[-i,])$residuals)
 res12=as.matrix(lm(X1[-i,]~X2[-i,])$residuals)
 
 
 
 m=plsr(res12~resY2,ncomp=1)
 
 
 A=Y[-i,]%*%t(t(m$projection))
 
 ss=lm(A~X1[-i,]+X2[-i,])
 
 
 auxx=Y[-i,]-X1[-i,]%*%ss$coefficients[2:21]%*%(t(m$projection))
 
 auxx1=lm(auxx~X2[-i,])$coefficients[2:3,]
 
 
 
 fit_prpls[i,,1]=colMeans(Y[-i,])+(X1[i,]-colMeans(X1[-i,]))%*%ss$coefficients[2:21]%*%(t(m$projection))+(X2[i,]-colMeans(X2[-i,]))%*%(auxx1)
 error_prpls[i,,1]=(Y[i,]-fit_prpls[i,,1])^2
 
}


### PPLS for the rest of the values for d

for (d in 2:ncol(Y)){
 for (i in 1:nrow(X1)){
   
   resY2=as.matrix(lm(Y[-i,]~X2[-i,])$residuals)
   res12=as.matrix(lm(X1[-i,]~X2[-i,])$residuals)
   
   m=plsr(res12~resY2,ncomp=d)
   
   A=Y[-i,]%*%t(t(m$projection))
   
   ss=lm(A~X1[-i,]+X2[-i,])
   
   auxx=Y[-i,]-X1[-i,]%*%ss$coefficients[2:21,1:d]%*%(t(m$projection))
   
   auxx1=lm(auxx~X2[-i,])$coefficients[2:3,]
   
   
   
   fit_prpls[i,,d]=colMeans(Y[-i,])+(X1[i,]-colMeans(X1[-i,]))%*%ss$coefficients[2:21,1:d]%*%(t(m$projection))+(X2[i,]-colMeans(X2[-i,]))%*%(auxx1)
   error_prpls[i,,d]=(Y[i,]-fit_prpls[i,,d])^2
   
 }
}


p.ols=NULL
p.pls=NULL
p.env=NULL

for (d in 1:ncol(Y)){
 p.env[d]=mean(sqrt(apply(error_penv[,,d],1,mean)))
 p.pls[d]=mean(sqrt(apply(error_prpls[,,d],1,mean)))
}




# Table 6.2 columns 5,6

table6_2[2,4] = which.min(p.pls)
table6_2[3,4] = p.pls[which.min(p.pls)]
table6_2[2,5] = which.min(p.env)
table6_2[3,5] = p.env[which.min(p.env)]

table6_2

File: examples/chapter6/table6-2.py
import warnings
warnings.filterwarnings("ignore")

import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.cross_decomposition import PLSRegression
import pandas as pd
from tqdm import tqdm
from tabulate import tabulate

np.random.seed(5)

# PPLS Partial OLS and ENV Partial table 6.2

# Columns 2,3,4

df = pd.read_csv("data/chapter6/Process_data_Skagerberg1992 - Sheet1.csv")

# For reduction
X1 = df.iloc[:, :20].to_numpy()

# No reduction
X2 = df.iloc[:, 20:22].to_numpy()

Y = np.log(df.iloc[:, 22:28].to_numpy())

Y = Y @ (np.diag(np.diag(np.cov(Y, ddof=1,rowvar=False)) ** (-1/2)))

# df = df.assign(Y=Y)

aux = df.to_numpy()

Dmax = 18

n = len(df)

pls_predict = np.zeros((n, Dmax, 6))
ols_predict = np.zeros((n, Dmax, 6))
# TODO no envelope module in python yet
# env_predict = np.zeros((n, Dmax, 6))

# PPLS partial OLS and ENV partial

for d in tqdm(range(Dmax), desc="Training pls, ols, and env models"):
    for i in range(n):
        # Training
        aux1 = np.delete(aux, i, 0)
        aux1_Y = np.delete(Y, i, 0)

        # Testing
        aux2 = aux[i]
        aux2_Y = Y[i]

        # Matrix X to reduce
        AA = aux1[:, :20]

        # print((aux1).shape, (aux1[df.columns.get_loc("S"), :]).shape)
        
        # Regression of the continuous X that we do not reduce
        aux1_Tw_S = np.stack((aux1[:, df.columns.get_loc("Tw")], aux1[:, df.columns.get_loc("S")])).T
        A = LinearRegression()
        A.fit(X=aux1_Tw_S, y=AA)

        # Save residuals
        aux1_SS = AA - A.predict(aux1_Tw_S)

        # Predict on new data
        aux2_Tw_S = np.stack((aux2[df.columns.get_loc("Tw")], aux2[df.columns.get_loc("S")])).reshape(-1,1).T
        A_1 = A.predict(aux2_Tw_S)

        # Center response in training
        aux1_ynew = aux1_Y - aux1_Y.mean(axis=0)

        # Do PLS on the training residuals
        m = PLSRegression(n_components=d+1, scale=False)
        m.fit(X=aux1_SS, Y=aux1_ynew)

        # Perform regression on non-reduced variables
        j = LinearRegression()
        j.fit(X=aux1_Tw_S, y=aux1_ynew)

        # Predict on the new datapoint from non-reduced model
        j_2 = j.predict(aux2_Tw_S)

        # Final prediction on new data

        yhat = aux1_Y.mean(axis=0) + j_2 + (aux2[0:20] - A_1) @ m.coef_.T

        pls_predict[i, d] = yhat
        
        aux1_a_predictors = np.delete(aux1, [df.columns.get_loc("Mw_inverse"), df.columns.get_loc("LCB"), df.columns.get_loc("SCB"), df.columns.get_loc("VNL"), df.columns.get_loc("VND")], 1)
        a = LinearRegression()
        a.fit(X=aux1_a_predictors, y=aux1_Y)
        
        aux2_a_predictors = np.delete(aux2, [df.columns.get_loc("Mw_inverse"), df.columns.get_loc("LCB"), df.columns.get_loc("SCB"), df.columns.get_loc("VNL"), df.columns.get_loc("VND")], 0).reshape(-1, 1).T
        ols_predict[i,d] = a.predict(aux2_a_predictors)

        # TODO No envelopes package
        # env_predict[i,d] = ...

p_ols = np.zeros((Dmax,))
p_pls = np.zeros((Dmax,))
# TODO No envelopes package
# p_env = np.zeros((Dmax,))

for d in range(Dmax):
    p_ols[d] = np.mean(np.sqrt(((Y - ols_predict[:,d])**2).mean(axis=1)))
    p_pls[d] = np.mean(np.sqrt(((Y - pls_predict[:,d])**2).mean(axis=1)))
    # TODO No envelopes package
    # p_env[d] = np.mean(np.sqrt(((Y - env_predict[:,d])**2).mean(axis=1)))

table6_2 = {
    "Method": ['OLS', 'PLS (P.Predictor)', 'Envelope (P.Predictor)', 'PLS (P.Response)', 'Envelope (P.Response)'],
    "No. components": ["not implemented"]*5,
    "predRMSE": ["not implemented"]*5
}

table6_2["No. components"][0] = "p1 = 20"
table6_2["predRMSE"][0] = p_ols[0]
table6_2["No. components"][1] = np.argmin(p_pls)+1
table6_2["predRMSE"][1] = np.min(p_pls)
# TODO No envelopes package
# table6_2["No. components"][2] = np.argmin(p_env)+!
# table6_2["predRMSE"][2] = np.min(p_env)


# Columns 5 and 6

# TODO No envelopes package
# fit_penv = np.zeros((X1.shape[0], Y.shape[1], X1.shape[1]))
# error_penv = np.zeros((X1.shape[0], Y.shape[1], X1.shape[1]))

# for d in range(Y.shape[1]):
#     for i in range(X1.shape[0]):
#         # Do envelope stuff


fit_prpls = np.zeros((X1.shape[0], Y.shape[1], X1.shape[1]))
error_prpls = np.zeros((X1.shape[0], Y.shape[1], X1.shape[1]))

# PPLS for d=1 we do it separately

for i in tqdm(range(X1.shape[0]), desc="Fitting PPLS for d=1"):

    Ythis = np.delete(Y, i, 0)
    X1this = np.delete(X1, i, 0)
    X2this = np.delete(X2, i, 0)

    lm = LinearRegression()
    lm.fit(X=X2this, y=Ythis)
    resY2 = Ythis - lm.predict(X2this)

    lm.fit(X=X2this, y=X1this)
    res12 = X1this - lm.predict(X2this)

    m = PLSRegression(n_components=1, scale=False)
    m.fit(X=resY2, Y=res12)

    A = Ythis @ (-m.x_rotations_)

    ss_X = np.hstack((X1this, X2this))
    ss = LinearRegression()
    ss.fit(X=ss_X, y=A)

    auxx = Ythis - X1this @ ss.coef_.T[0:20] @ (-m.x_rotations_.T)

    auxx1 = LinearRegression()
    auxx1.fit(X=X2this, y=auxx)
    auxx1 = auxx1.coef_

    fit_prpls[i, :, 0] = Ythis.mean(axis=0) + (X1[i] - X1this.mean(axis=0)) @ ss.coef_.T[0:20] @ (-m.x_rotations_.T) + (X2[i] - X2this.mean(axis=0)) @ auxx1.T

    error_prpls[i,:,0] = (Y[i] - fit_prpls[i,:,0])**2


# PPLS for the rest of the values for d
for d in tqdm(range(1, Y.shape[1]), desc="Fitting PPLS for 1<d<=ncol(Y)"):
    for i in range(X1.shape[0]):
        Ythis = np.delete(Y, i, 0)
        X1this = np.delete(X1, i, 0)
        X2this = np.delete(X2, i, 0)

        lm = LinearRegression()
        lm.fit(X=X2this, y=Ythis)
        resY2 = Ythis - lm.predict(X2this)

        lm.fit(X=X2this, y=X1this)
        res12 = X1this - lm.predict(X2this)

        m = PLSRegression(n_components=d+1, scale=False)
        m.fit(X=resY2, Y=res12)

        A = Ythis @ (-m.x_rotations_)

        ss_X = np.hstack((X1this, X2this))
        ss = LinearRegression()
        ss.fit(X=ss_X, y=A)

        auxx = Ythis - X1this @ ss.coef_.T[0:20] @ (-m.x_rotations_.T)

        auxx1 = LinearRegression()
        auxx1.fit(X=X2this, y=auxx)
        auxx1 = auxx1.coef_

        fit_prpls[i, :, d] = Ythis.mean(axis=0) + (X1[i] - X1this.mean(axis=0)) @ ss.coef_.T[0:20] @ (-m.x_rotations_.T) + (X2[i] - X2this.mean(axis=0)) @ auxx1.T

        error_prpls[i,:,d] = (Y[i] - fit_prpls[i,:,d])**2

# print(error_prpls)

p_pls = np.zeros((Y.shape[1],))
# TODO No envelopes package
# p_env = np.zeros((Y.shape[1],))

for d in range(Y.shape[1]):
    p_pls[d] = np.mean(np.sqrt(error_prpls[:,:,d].mean(axis=1)))
    # TODO No envelopes package
    # p_env[d] = np.mean(np.sqrt(error_penv[:,:,d].mean(axis=1)))

table6_2["No. components"][3] = np.argmin(p_pls)+1
table6_2["predRMSE"][3] = np.min(p_pls)
# TODO No envelopes package
# table6_2["No. components"][4] = np.argmin(p_env)+1
# table6_2["predRMSE"][4] = np.min(p_env)

print()
print("Table 6.2 (transposed):")
print(tabulate(table6_2, headers=table6_2.keys()))
print()